brushfire-core/src/main/scala/com/stripe/brushfire/Tree.scala

```diff
@@ -31,6 +32,107 @@ case class Tree[K, V, T](root: Node[K, V, T]) {
     }
   }
 
+  /**
+   * Prune a tree to minimize validation error.
+   * Replaces a split with a leaf when it would have a lower error than the child leaves.
+   *
+   * @param validationData
+   * @param start
+   * @param voter
+   * @param error
+   * @param weightMonoid
+   * @param errorOrdering
+   * @tparam P
+   * @tparam E
+   * @return
+   */
+  def prune[P, E](validationData: Map[Int, T], start: Option[Node[K, V, T]] = None, voter: Voter[T, P], error: Error[T, P, E])(implicit weightMonoid: Monoid[T], errorOrdering: Ordering[E]): Tree[K, V, T] = {
+    val begin: Node[K, V, T] = start.getOrElse(this.root) // Use root node as default root.
+    Tree(pruneNode(validationData, begin, voter, error))
+      .growByLeafIndex { i => Nil } //this will renumber all the leaves
+  }
+
+  /**
+   * Prune a tree to minimize validation error, starting from given root node.
+   *
+   * This method recursively traverses up the tree from the root, branching on splits, until it finds leaves, then
+   * goes back down the tree combining leaves when such a combination would reduce validation error.
+   *
+   * @param validationData
+   * @param start
+   * @param voter
+   * @param error
+   * @param weightMonoid
+   * @param errorOrdering
+   * @tparam P
+   * @tparam E
+   * @return
+   */
+  def pruneNode[P, E](validationData: Map[Int, T], start: Node[K, V, T], voter: Voter[T, P], error: Error[T, P, E])(implicit weightMonoid: Monoid[T], errorOrdering: Ordering[E]): Node[K, V, T] = {
+    val vData = mutable.Map() ++ validationData // Create a mutable Map to allow merging of validation data as a side effect of the pruneLevel method.
```

Comment:
instead of a mutable `vData` here, could you have it return a tuple with the new node and also the new validationData? That would be more idiomatic.